File: 0567-permutation-in-string/0567-permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        if(s1.length() > s2.length()){
            return false;
        }
        int[] s1count = new int[26];
        int[] s2count = new int[26];
        for(int i = 0 ; i < s1.length() ; i++){
            s1count[s1.charAt(i) - 'a'] += 1;
            s2count[s2.charAt(i) - 'a'] += 1;
        }
        int match = 0;
        for(int i = 0 ; i < 26 ; i++){
            if(s1count[i] == s2count[i]){
                match += 1;
            }
        }
        int l = 0;
        for(int r = s1.length() ; r < s2.length() ; r++){
            if(match == 26){
                return true;
            }

            int index = s2.charAt(r) - 'a';
            s2count[index] += 1;
            if(s2count[index] == s1count[index]){
                match += 1;
            }else if(s2count[index] == s1count[index] + 1){
                match -= 1;
            }

            index = s2.charAt(l) - 'a';
            s2count[index] -= 1;
            if(s2count[index] == s1count[index]){
                match += 1;
            }else if(s2count[index] == s1count[index] - 1){
                match -= 1;
            }

            l += 1;
        }
        return match == 26;
    }
}

File: 0567-permutation-in-string/0567-permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2) :
            return False

        s1count , s2count = [0] * 26 , [0] * 26
        for i in range(len(s1)):
            s1count[ord(s1[i]) - ord("a")] += 1
            s2count[ord(s2[i]) - ord("a")] += 1

        if s1count == s2count:
            return True
        
        l = 0
        for r in range(len(s1), len(s2)):
            # 將右邊界的字符納入窗口
            s2count[ord(s2[r]) - ord("a")] += 1
            # 將左邊界的字符移出窗口
            s2count[ord(s2[l]) - ord("a")] -= 1
            # 如果窗口大小等於 p 的長度，檢查是否找到了一個字母排列
            if s1count == s2count:
                return True
                
            
            
            l += 1
        return False


        